src/libtoast/src/toast_tod_filter.cpp

```diff
@@ -52,108 +58,228 @@ void toast::filter_polynomial(int64_t order, size_t n, uint8_t * flags,
+                size_t offset = isignal * ngood;
+                double * signal = signals[isignal] + start;
+                for (int64_t i = 0; i < scanlen; ++i) {
+                    if (flags[start + i] != 0) continue;
+                    masked_signals[offset++] = signal[i];
                 }
-                continue;
             }
 
-            // Symmetrize for dgemv later on (workaround for dtrmv issue)
+            toast::AlignedVector <double> proj(norder * nsignal);
 
-            for (int row = 0; row < norder; ++row) {
-                for (int col = row + 1; col < norder; ++col) {
-                    cov[row * norder + col] = cov[col * norder + row];
-                }
-            }
+            toast::lapack_gemm(&trans, &notrans, &norder, &nsignal, &ngood,
+                               &fone, masked_templates.data(), &ngood,
+                               masked_signals.data(), &ngood,
+                               &fzero, proj.data(), &norder);
 
-            // Filter every signal
+            // Symmetrize the covariance matrix, dgells is written for
+            // generic matrices
 
-            toast::AlignedVector <double> proj(norder);
-            toast::AlignedVector <double> coeff(norder);
-            toast::AlignedVector <double> noise(scanlen);
+            for (size_t row = 0; row < norder; ++row) {
+                for (size_t col = row + 1; col < norder; ++col) {
+                    invcov[col + row * norder] = invcov[row + col * norder];
+                }
+            }
 
-            for (size_t isignal = 0; isignal < nsignal; ++isignal) {
-                double * signal = signals[isignal] + start;
+            // Fit the templates against the data.  DGELSS uses SVD to
+            // minimize the norm of the difference and the solution
+            // vector.
+
+            toast::AlignedVector <double> singular_values(norder);
+            int rank, info;
+            double rcond_limit = 1e-3;
+            int lwork = std::max(10 * (norder + nsignal), 1000000);
+            toast::AlignedVector <double> work(lwork);
+
+            // DGELSS will overwrite proj with the fitting
+            // coefficients.  invcov is overwritten with
+            // singular vectors.
+            toast::lapack_dgelss(&norder, &norder, &nsignal,
+                                 invcov.data(), &norder,
+                                 proj.data(), &norder,
+                                 singular_values.data(), &rcond_limit,
+                                 &rank, work.data(), &lwork, &info);
+
+            for (int iorder = 0; iorder < norder; ++iorder) {
+                double * temp = &full_templates[iorder * scanlen];
+                for (int isignal = 0; isignal < nsignal; ++isignal) {
+                    double * signal = &signals[isignal][start];
+                    double amp = proj[iorder + isignal * norder];
+                    #pragma omp simd
```

Comment:
Here is a case where signal points to the `signals` variable which is passed into the function and of unknown alignment.  The temp variable points to `full_templates` which is allocated in aligned memory.  There is a helper function for checking alignment:
```
if (toast::is_aligned(signals)) ...
```
See the healpix or qarray source files for examples.

Reply:
Thanks. Done.